File: fabric-audit-agent/core/sanitize.js
const SAFE_STRING_KEYS = new Set(['sku', 'status', 'time']); // enum-like, not identifying

/** Strip identifiers from one evidence object: keep numbers/booleans + safe enums; arrays -> count; drop other strings. */
export function sanitizeEvidence(evidence = {}) {
  if (evidence.sensitive === true || evidence.sensitivityLabel) return { redacted: true };
  const out = {};
  for (const [k, v] of Object.entries(evidence)) {
    if (typeof v === 'number' || typeof v === 'boolean') out[k] = v;
    else if (Array.isArray(v)) out[`${k}Count`] = v.length;
    else if (typeof v === 'string' && SAFE_STRING_KEYS.has(k)) out[k] = v;
    // else: drop (dataset names, sources, free text, timestamps)
  }
  return out;
}

/**
 * Build the external-safe payload from flags: only an index, the flag type, and
 * sanitized numeric evidence. No resource, no `what`, no names.
 * @param {object[]} flags
 * @returns {Array<{id:number, type:string, evidence:object}>}
 */
export function sanitize(flags) {
  return flags.map((f, i) => ({ id: i, type: f.type, evidence: sanitizeEvidence(f.evidence) }));
}

File: fabric-audit-agent-py/fabric_audit_agent/sanitize.py
"""Sanitize evidence before it leaves the tenant (e.g. to an LLM). Port of ``core/sanitize.js``.

Keep numbers/booleans + a few safe enum strings; arrays -> count; drop identifying strings.
"""
_SAFE_STRING_KEYS = {"sku", "status", "time"}   # enum-like, not identifying


def sanitize_evidence(evidence=None):
    evidence = evidence or {}
    if evidence.get("sensitive") is True or evidence.get("sensitivityLabel"):
        return {"redacted": True}
    out = {}
    for k, v in evidence.items():
        if isinstance(v, (int, float)):        # numbers + booleans (bool is an int subclass, mirrors JS number||boolean)
            out[k] = v
        elif isinstance(v, list):
            out[f"{k}Count"] = len(v)
        elif isinstance(v, str) and k in _SAFE_STRING_KEYS:
            out[k] = v
        # else: drop (dataset names, sources, free text, timestamps)
    return out


def sanitize(flags):
    """External-safe payload: index + flag type + sanitized numeric evidence only."""
    return [{"id": i, "type": f.get("type"), "evidence": sanitize_evidence(f.get("evidence"))} for i, f in enumerate(flags)]
